Battle UI: Don't use `switchposition` controls in Gen 7

Slayer95 opened the pull request:
Forced slot switching

Zarel commented on js/client-battle.js:
```diff
@@ -379,8 +379,8 @@
-						this.choice.freedomDegrees = Math.max(freedomDegrees, 0);
-						this.choice.canSwitch = faintedLength - this.choice.freedomDegrees;
+						this.choice.canSwitch = faintedLength - Math.max(freedomDegrees, 0);
+						this.choice.freedomDegrees = this.battle.gen >= 7 ? 0 : Math.max(freedomDegrees, 0);
```
I'm confused. Shouldn't `freedomDegrees` be sent from the server?

Slayer95 replied:
That makes more sense. Though are you fine with that key name?

> In statistics, the number of degrees of freedom is the number of values in the final calculation of a statistic that are free to vary

Zarel replied:
I really dislike the key name.

Zarel replied:
Maybe the flag should be `canChooseSwitchLocation`.

Zarel commented:
I think I'd rather have this as a part of `request` than to have it be inferred from the gen.

In general I think data about what sort of choices are legal should be sent explicitly through the request, rather than needing to be inferred from anything else.

Slayer95 commented:
@Zarel , actually this should be merged as is before implementing the flag in the server for compatibility.

Zarel commented:
I don't see a check for the flag. Should I wait for that?

Slayer95 commented:
There is an issue with ``canChooseSwitchLocation``.

Strictly speaking, the cartridge regularly requests the player to choose switch locations when 2 Pokémon faint. Now, because the issued commands are commutative, we could avoid requesting for locations unless the player could opt for leaving an amount of slots empty (equal to ``freedomDegrees``).

In Gen 7, the player can no longer opt for leaving slots empty. However, the cartridge still requests for switch locations if 2 Pokémon faint at the same time and can both be replaced. Following that, it would be a shame to have a ``canChooseSwitchLocation`` flag in the protocol which isn't accurate to the in-game behavior.

Slayer95 commented:
Now, that issue would be non-existent if we followed the in-game UI to the letter. However, I believe that would add undeserved clunkyness.

Zarel commented:
Our UI already doesn't match cartridge though, right? Feel free to either fix or not depending on your preference.

Slayer95 commented:
> There is an issue with canChooseSwitchLocation.

I'd rather have a different name, and punt the issue of matching cartridge regarding requests for switch locations for the future.

Slayer95 commented:
Alright. I made up my mind on an action path.
1. Use ``canChooseSwitchLocation``
2. Keep our divergences in UI from cartridge.
3. If we ever decide to change point 2, it can be easily toggled server-side.
4. Hardcore mode can be made to override ``canChooseSwitchLocation``, but user demand should be probed.

With that stated and if there are no objections, this is ready !

Slayer95 commented:
Server PR: https://github.com/Zarel/Pokemon-Showdown/pull/5390

Zarel commented:
Maybe `forcedSwitchLocations`?

Zarel commented:
Specifically, if switch locations are chosen in double-faints normally, the flag isn't so much about "you can choose the switch location" so much as "you're forced to switch into the leftmost available slot".

Slayer95 commented:
> If we ever decide to change point 2, it can be easily toggled server-side.

I believe there is value in that. Even if we don't change the UI, a VGC-dedicated server could opt for doing so.

On the other hand, the semantics of ``forcedSwitchLocations`` imply that it's equivalent to ``battle.gen === 7```, so the UI would be determined by the client. Or it could just be the negation of ``canChooseSwitchLocations`` ? .....

Sigh, this bikeshedding is getting tiresome.

Zarel commented:
Hey, protocol changes are important enough that I personally don't consider them bikeshedding. They're like the one thing we have to support long-term.

Slayer95 commented:
Do let me know if you reach a final plan though.

Zarel commented:
Anyway, it doesn't make sense for accurate simulation of something UI-only to be done server-side. It should be client-side, and the client-side option (hardcore mode) already exists.

Zarel commented:
So yes, `forcedSwitchLocations` is supposed to be equal to `battle.gen >= 7`, and that's still my preferred approach. Feel free to suggest other names, but you've yet to present a convincing argument otherwise for mechanics.

And if at any point you want to stop arguing and decide it all yourself, you have my approval for that, too.

Slayer95 commented:
Well, that's fine too. I'll check the implementation details to account for that naming later.

Slayer95 committed:
Battle UI: Update switchposition usage

- Implement Gen 7 forced slot switching.
- Support strict in-game UI through hardcore mode setting.
